src/appengine/private/components/upload-testcase/upload-form-simplified.html

```diff
@@ -0,0 +1,391 @@
+          'js':
+            {
+              'Linux': ['linux_asan_d8', null],
+              'Windows': ['windows_asan_d8', null],
+              'Mac': ['mac_asan_d8', null],
+              'Android': ["No applicable job available! :( Please do not select this option.", null]
+            },
+          'pdf':
+            {
+              'Linux': ['libfuzzer_pdfium_asan', 'pdfium_xfa_fuzzer'],
+              'Windows': ['libfuzzer_pdfium_asan', 'pdfium_xfa_fuzzer'],
+              'Mac': ['libfuzzer_pdfium_asan', 'pdfium_xfa_fuzzer'],
+              'Android': ['libfuzzer_pdfium_asan', 'pdfium_xfa_fuzzer']
+            }
+        }
+
+        let fileTypeKey;
+        if (isArchive) {
+          fileTypeKey = fileType;
+        } else {
+          fileTypeKey = file.split('.').pop();
+        }
+        if (jobMapping[fileTypeKey] && jobMapping[fileTypeKey][platform]) {
+          let runParams = jobMapping[fileTypeKey][platform];
+          this.uploadParams.job = runParams[0];
+          this.uploadParams.target = runParams[2];
+        } else {
+          // We should never get here. Throw an error and perhaps work on displaying it if need be, later.
+          throw 'The combo of platform: ' + platform + ' and file: ' + file + 'and maybe the fileType was unexpected!'
+        }
+        return true;
+      }
+
+      computeGcsBody() {
+        var uploadData = new FormData();
+        uploadData.append('bucket', this.fieldValues.uploadInfo.bucket);
+        uploadData.append('key', this.fieldValues.uploadInfo.key);
+        uploadData.append('GoogleAccessId', this.fieldValues.uploadInfo.google_access_id);
+        uploadData.append('policy', this.fieldValues.uploadInfo.policy);
+        uploadData.append('signature', this.fieldValues.uploadInfo.signature);
+
+        let file = this.$.file.inputElement.inputElement.files[0];
+        if (file) {
+          uploadData.append('x-goog-meta-filename', file.name);
+          uploadData.append('file', file);
+        } else {
+          this.$.ajaxFinish.generateRequest();
+          return;
+        }
+
+        this.$.dialog.$.ajax.body = uploadData;
+      }
+
+      computeFinishBody() {
+        let formData = new FormData();
+        for (let key in this.uploadParams) {
+          formData.append(key, this.uploadParams[key]);
+        }
+
+        formData.append('csrf_token', this.fieldValues.csrfToken);
+        if (this.uploadParams.file) {
+          formData.append('upload_key', this.fieldValues.uploadInfo.key);
+        }
+        return formData;
+      }
+
+      handleGcsResponseAndError(ev) {
+        // Finished uploading to GCS, post the remainder of the data to
+        // ClusterFuzz.
+        // Errors are ignored here also. Because the upload is cross-site, we
+        // defer to the server to check that upload succeeded.
+        this.$.ajaxFinish.generateRequest();
+      }
+
+      handleResponse(ev) {
+        this.set('uploadParams.job', uploadParams.job);
```

Comment:
The flow of data is a little confusing, and it seems like having this here fixes it inadvertently, but it's not the right place for it.

This should instead be called whenever the `job` variable itself is set (i.e. line 288 in your current revision), but following the thread a little, it's a bit confusing why `updateJobFromPlatformAndFile` is only called from `showWarning`. 

The way this template framework works is more declarative and event driven -- it seems like you're trying to make it write things in an imperative way which leads to all of these issues you're encountering.  You can think of the arguments to these functions here as "dependencies", rather than inputs -- if one of the dependencies changes (i.e. a parameter in a textbox), then the function will be re-called to recalculate the output. 

I'd suggest moving `updateJobFromPlatformAndFile` into `getJobType` inline itself, move the `this.set` call there, and change `showWarning` to accept a job type name instead. 

Then your showWarning call will look something like:

```
<template is="dom-if" if="[[showWarning(uploadParams.job)]]">
```

This will get updated when the frontend framework detects that `uploadParams.job` has changed (via the `this.set` call in `getJobType` which is needed because `uploadParams` is Javascript object). `getJobType` itself has dependencies on `uploadParams.file, uploadParams.fileType, uploadParams.platform`, which may or may not need `this.set` calls in the right places to trigger. 

Does this make sense?

Reply:
yesss totally agreed!! thank you!!